File: ios/collectMissingKeys.swift
//
//  collectMissingKeys.swift
//  openland
//
//  Created by Steve Kite on 5/13/19.
//  Copyright © 2019 Openland. All rights reserved.
//

import Foundation
import SwiftyJSON

fileprivate func collectMissingKeys(cacheKey: String, selectors: [Selector], store: RecordStore, variables: JSON) -> Set<String> {
  if !store.isInMemory(key: cacheKey) {
    return Set([cacheKey])
  }
  let value = store.read(key: cacheKey)
  var res = Set<String>()
  for f in selectors {
    if f is Selector.Field {
      let fl = f as! Selector.Field
      let key = selectorKey(name: fl.name, arguments: fl.arguments, variables: variables)
      if value.fields[key] != nil {
        res = res.union(collectMissingKeys(value: value.fields[key]!, type: fl.type, store: store, variables: variables))
      }
    } else if f is Selector.TypeCondition {
      let tc = f as! Selector.TypeCondition
      if value.fields["__typename"] == RecordValue.StringValue(value: tc.name) {
        res = res.union(collectMissingKeys(cacheKey: cacheKey, selectors: tc.type.selectors, store: store, variables: variables))
      }
    } else if f is Selector.Fragment {
      let fr = f as! Selector.Fragment
      res = res.union(collectMissingKeys(cacheKey: cacheKey, selectors: fr.type.selectors, store: store, variables: variables))
    } else {
      fatalError()
    }
  }
  return res
}

fileprivate func collectMissingKeys(value: RecordValue, type:OutputType, store:RecordStore, variables: JSON) -> Set<String> {
  if type is OutputType.Scalar {
    return Set()
  } else if type is OutputType.NotNull {
    if value is RecordValue.NullValue {
      return Set()
    } else {
      return collectMissingKeys(value: value, type: (type as! OutputType.NotNull).inner, store: store, variables: variables)
    }
  } else if type is OutputType.List {
    if value is RecordValue.ListValue {
      let ls = value as! RecordValue.ListValue
      let lt = (type as! OutputType.List)
      var res = Set<String>()
      for i in ls.items {
        res = res.union(collectMissingKeys(value: i, type: lt.inner, store: store, variables: variables))
      }
      return res
    } else {
      return Set()
    }
  } else if type is OutputType.Object {
    if value is RecordValue.ReferenceValue {
      let ref = (value as! RecordValue.ReferenceValue)
      let tp = (type as! OutputType.Object)
      return collectMissingKeys(cacheKey: ref.key, selectors: tp.selectors, store: store, variables: variables)
    } else {
      return Set()
    }
  }
  
  return Set()
}

func collectMissingKeys(cacheKey: String, store: RecordStore, type: OutputType.Object, variables: JSON) -> Set<String> {
  return collectMissingKeys(cacheKey: cacheKey, selectors: type.selectors, store: store, variables: variables)
}

func collectMissingKeysRoot(cacheKey: String, store: RecordStore, type: OutputType.Object, variables: JSON) -> Set<String> {
  var res = Set<String>()
  for f in type.selectors {
    if !(f is Selector.Field) {
      fatalError("Root query can't contain fragments")
    }
    
    let sf = f as! Selector.Field
    let key = selectorKey(name: sf.name, arguments: sf.arguments, variables: variables)
    let refId = "\(cacheKey).$ref.\(key)"
    if !store.isInMemory(key: refId) {
      res.insert(refId)
    } else {
      let value = store.read(key: refId)
      let ex = value.fields["data"]
      if ex != nil {
        for s in collectMissingKeys(value: ex!, type: sf.type, store: store, variables: variables) {
          res.insert(s)
        }
      }
    }
  }
  return res
}

File: android/app/src/main/java/com/openland/spacex/store/collectMissingKeys.kt
package com.openland.spacex.store

import com.openland.spacex.OutputType
import com.openland.spacex.Selector
import org.json.JSONObject

/**
 * Collect Missing Keys from selectors
 * @param cacheKey Root Cache Key
 * @param store Store
 * @param selectors Selectors
 * @param arguments Query arguments
 * @return set of missing keys
 */
private fun collectMissingKeys(cacheKey: String, store: RecordStore, selectors: List<Selector>, arguments: JSONObject): Set<String> {
    if (!store.isInMemory(cacheKey)) {
        return setOf(cacheKey)
    }
    val record = store.read(cacheKey)

    val res = mutableSetOf<String>()
    for (f in selectors) {
        if (f is Selector.Field) {
            val key = selectorKey(f.name, f.arguments, arguments)
            if (record.fields.containsKey(key)) {
                res += collectMissingKeys(record.fields[key]!!, f.type, store, arguments)
            }
        } else if (f is Selector.TypeCondition) {
            if (record.fields["__typename"] == RecordValue.String(f.type)) {
                res += collectMissingKeys(cacheKey, store, f.fragment.selectors, arguments)
            }
        } else if (f is Selector.Fragment) {
            res += collectMissingKeys(cacheKey, store, f.fragment.selectors, arguments)
        } else {
            error("Unsupported selector")
        }
    }
    return res
}

/**
 * Collect Missing Keys from a value
 * @param value Record Value
 * @param type Expected Value Type
 * @param store Store
 * @param arguments Query arguments
 * @return set of missing keys
 */
private fun collectMissingKeys(value: RecordValue, type: OutputType, store: RecordStore, arguments: JSONObject): Set<String> {
    if (type is OutputType.Scalar) {
        return emptySet()
    } else if (type is OutputType.NotNull) {
        return if (value == RecordValue.Null) {
            emptySet()
        } else {
            collectMissingKeys(value, type.inner, store, arguments)
        }
    } else if (type is OutputType.List) {
        return if (value == RecordValue.Null) {
            emptySet()
        } else {
            if (value is RecordValue.List) {
                val res = mutableSetOf<String>()
                val mapped = value.items.map { collectMissingKeys(it, type.inner, store, arguments) }
                for (m in mapped) {
                    res += m
                }
                res
            } else {
                error("Invalid record value")
            }
        }
    } else if (type is OutputType.Object) {
        return if (value == RecordValue.Null) {
            emptySet()
        } else {
            if (value is RecordValue.Reference) {
                collectMissingKeys(value.key, store, type.selectors, arguments)
            } else {
                error("Invalid record value")
            }
        }
    }
    return emptySet()
}

/**
 * Collect object keys that are needed for query and not present in RecordStore
 * @param rootCacheKey Root Cache Key (usually ROOT_QUERY)
 * @param store Store
 * @param type Expected type
 * @param arguments Query Arguments
 * @return set of missing keys
 */
fun collectMissingKeysRoot(rootCacheKey: String, store: RecordStore, type: OutputType.Object, arguments: JSONObject): Set<String> {
    val res = mutableSetOf<String>()
    for (f in type.selectors) {
        if (f !is Selector.Field) {
            error("Root query cant't contain fragments")
        }
        val key = selectorKey(f.name, f.arguments, arguments)
        val refId = "$rootCacheKey.\$ref.$key"
        if (!store.isInMemory(refId)) {
            res += refId
        } else {
            val value = store.read(refId)
            val ex = value.fields["data"]
            if (ex != null) {
                res += collectMissingKeys(ex, f.type, store, arguments)
            }
        }
    }
    return res
}